streamer/src/sendmmsg.rs

```diff
@@ -161,27 +178,41 @@ fn sendmmsg_retry(sock: &UdpSocket, hdrs: &mut [mmsghdr]) -> Result<(), SendPkts
 }
 
 #[cfg(target_os = "linux")]
-pub fn batch_send<S, T>(sock: &UdpSocket, packets: &[(T, S)]) -> Result<(), SendPktsError>
+const MAX_IOV: usize = libc::UIO_MAXIOV as usize;
+
+#[cfg(target_os = "linux")]
+pub fn batch_send_max_iov<S, T>(sock: &UdpSocket, packets: &[(T, S)]) -> Result<(), SendPktsError>
 where
     S: Borrow<SocketAddr>,
     T: AsRef<[u8]>,
 {
-    let size = packets.len();
-    let mut iovs = vec![MaybeUninit::uninit(); size];
-    let mut addrs = vec![MaybeUninit::zeroed(); size];
-    let mut hdrs = vec![MaybeUninit::uninit(); size];
+    assert!(packets.len() <= MAX_IOV);
+
+    let mut iovs = [MaybeUninit::uninit(); MAX_IOV];
+    let mut addrs = [MaybeUninit::uninit(); MAX_IOV];
+    let mut hdrs = [MaybeUninit::uninit(); MAX_IOV];
+
+    // izip! will iterate packets.len() times, leaving hdrs, iovs, and addrs initialized only up to packets.len()
     for ((pkt, dest), hdr, iov, addr) in izip!(packets, &mut hdrs, &mut iovs, &mut addrs) {
         mmsghdr_for_packet(pkt.as_ref(), dest.borrow(), iov, addr, hdr);
     }
-    // mmsghdr_for_packet() performs initialization so we can safely transmute
-    // the Vecs to their initialized counterparts
-    let _iovs = unsafe { mem::transmute::<Vec<MaybeUninit<iovec>>, Vec<iovec>>(iovs) };
-    let _addrs = unsafe {
-        mem::transmute::<Vec<MaybeUninit<sockaddr_storage>>, Vec<sockaddr_storage>>(addrs)
-    };
-    let mut hdrs = unsafe { mem::transmute::<Vec<MaybeUninit<mmsghdr>>, Vec<mmsghdr>>(hdrs) };
```

Comment:
The transmutes definitely need to go away, since we no longer know if the entire array will be initialized. @alessandrod and I had a conversation about similar type of problem here: https://github.com/anza-xyz/agave/pull/3348#discussion_r1820140698

The conclusion reached in that thread resulted in:
https://github.com/anza-xyz/agave/blob/3deac722ded065f47b36936d27480319f55dc029/streamer/src/recvmmsg.rs#L177-L190